### PR TITLE
8268148: unchecked warnings handle ? and ? extends Object differently

### PR DESCRIPTION
**Pull request by vicente-romero-oracle**
Please review this PR. Currently javac handles differently `?` and `? extends Object` when reporting unchecked warnings when according to the spec they should be handled the same way. This PR is fixing this.

TIA

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8268148](https://bugs.openjdk.java.net/browse/JDK-8268148): unchecked warnings handle ? and ? extends Object differently

### Reviewers
 * [Jan Lahoda](https://openjdk.java.net/census#jlahoda) (@lahodaj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5224/head:pull/5224` \
`$ git checkout pull/5224`

Update a local copy of the PR: \
`$ git checkout pull/5224` \
`$ git pull https://git.openjdk.java.net/jdk pull/5224/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5224`

View PR using the GUI difftool: \
`$ git pr show -t 5224`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5224.diff">https://git.openjdk.java.net/jdk/pull/5224.diff</a>

</details>

### CODE REVIEW

**Comment by vicente-romero-oracle** [comment]
ping

File: test/langtools/tools/javac/warnings/UnboundAndBoundByObjectTest.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8268148
+ * @summary unchecked warnings handle ? and ? extends Object differently
+ * @compile -Xlint:all UnboundAndBoundByObjectTest.java
```
Comment (lahodaj) [inline test/langtools/tools/javac/warnings/UnboundAndBoundByObjectTest.java:28]: Should there be a verification of the output? Possibly `-Werror`?

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java [hunk 2]
```diff
@@ -877,7 +877,9 @@ public boolean isExtendsBound() {
                 kind == UNBOUND;
         }
         public boolean isUnbound() {
-            return kind == UNBOUND;
+            // is it `?` or `? extends Object`?
+            return kind == UNBOUND ||
+                    (kind == EXTENDS && type.hasTag(CLASS) && ((ClassType)type).supertype_field == Type.noType);
```
Comment (lahodaj) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java:882]: For the test on whether `type` is j.l.Object, would something like this work better?
```suggestion
                    (kind == EXTENDS && type.tsym.flatName() == type.tsym.name.table.names.java_lang_Object);
```

Just an idea. (Sadly, we can't use `Symtab` here, that would be ideal.)

**Review by lahodaj — COMMENTED** [review]
Overall, looks reasonable. A few questions/comments inline.

**Review by lahodaj — APPROVED** [review]
Looks good, thanks!

**Comment by vicente-romero-oracle** [comment]
> Looks good, thanks!

thanks for the review

**Pull request closed without merging**